json/pickler/src/main/scala/sttp/tapir/json/pickler/Pickler.scala

```diff
@@ -0,0 +1,368 @@
+
+    if (!isValueCaseClass) {
+      '{ nonMirrorPickler[T] }
+    } else {
+
+      val field = tpe.typeSymbol.declaredFields.head
+      val fieldTpe = tpe.memberType(field)
+      fieldTpe.asType match
+        case '[f] =>
+          val basePickler = Expr.summon[Pickler[f]].getOrElse {
+            report.errorAndAbort(
+              s"Cannot summon Pickler for value class ${tpe.show}. Missing Pickler[${fieldTpe.show}] in implicit scope."
+            )
+          }
+          '{
+            val newSchema: Schema[T] = ${ basePickler }.schema.as[T]
+            new Pickler[T](
+              new TapirPickle[T] {
+                override lazy val writer = summonInline[Writer[f]].comap[T](
+                  // writing object of type T means writing T.field
+                  ccObj => ${ Select.unique(('ccObj).asTerm, field.name).asExprOf[f] }
+                )
+                // a reader of type f (field) will read it and wrap into value object using the consutructor of T
+                override lazy val reader = summonInline[Reader[f]]
+                  .map[T](fieldObj => ${ Apply(Select.unique(New(Inferred(tpe)), "<init>"), List(('fieldObj).asTerm)).asExprOf[T] })
+              },
+              newSchema
+            )
+          }
+    }
+
+  private inline def fromExistingSchemaAndRw[T](schema: Schema[T])(using ClassTag[T], Configuration, Mirror.Of[T]): Pickler[T] =
+    Pickler(
+      new TapirPickle[T] {
+        val rw: ReadWriter[T] = summonFrom {
+          case foundTapirRW: ReadWriter[T] =>
+            foundTapirRW
+          case foundUpickleDefaultRW: _root_.upickle.default.ReadWriter[T] => // there is BOTH schema and ReadWriter in scope
+            foundUpickleDefaultRW.asInstanceOf[ReadWriter[T]]
+          case _ =>
+            errorForType[T](
+              "Found implicit Schema[%s] but couldn't find a uPickle ReadWriter for this type. Either provide a ReadWriter, or remove the Schema from scope and let Pickler derive its own."
+            )
+            null
+        }
+        override lazy val reader = rw
+        override lazy val writer = rw
+      },
+      schema
+    )
+
+  private[pickler] inline def buildNewPickler[T: ClassTag](
+  )(using m: Mirror.Of[T], c: Configuration, subtypeDiscriminator: SubtypeDiscriminator[T]): Pickler[T] =
+    // The lazy modifier is necessary for preventing infinite recursion in the derived instance for recursive types such as Lst
+    lazy val childPicklers: Tuple.Map[m.MirroredElemTypes, Pickler] = summonChildPicklerInstances[T, m.MirroredElemTypes]
+    inline m match {
+      case p: Mirror.ProductOf[T] => picklerProduct(p, childPicklers)
+      case _: Mirror.SumOf[T] =>
+        val schema: Schema[T] =
+          inline if (isScalaEnum[T])
+            Schema.derivedEnumeration[T].defaultStringBased
+          else
+            Schema.derived[T]
```

Comment:
it's a bit suspicious that products & coproducts are handled differently ;) shouldn't we use `subtypeDiscriminator` when deriving coproduct schemas as well?

Reply:
This is indeed unclear. Looks like I'm using the subtype discriminator for both product and sum, while it's unnecessary for product. Additionaly, this method is called only in one place, with `DefaultSubtypeDiscriminator`, so it's also an unneeded parameter on this level.
We use discriminator only when it's "non-default", and it goes through another path, not `buildNewPickler`. This will hopefully become clearer when I remove the discriminator parameter from `buildNewPickler`. Just pushed.